File: python/aspectly_bridge/browser_bridge.py
from abc import ABC, abstractmethod


class BrowserBridge(ABC):
    """Abstraction for browser message passing.

    Implement this for any web view (WebKitGTK, a mock for testing, etc.).
    The Python equivalent of .NET's ``IBrowserBridge``.

    ``BridgeHost`` assigns ``on_message`` (a callable taking the message string)
    when it is constructed.
    """

    on_message = None

    @property
    @abstractmethod
    def is_ready(self):
        """Whether the browser is ready for communication."""

    @abstractmethod
    def execute_script(self, script):
        """Send a message to JavaScript by executing ``script``."""

    @abstractmethod
    def dispose(self):
        """Release resources and detach from the underlying web view."""


class BridgeLogger(ABC):
    """Simple logging interface for ``BridgeHost``."""

    @abstractmethod
    def debug(self, message):
        ...

    @abstractmethod
    def info(self, message):
        ...

    @abstractmethod
    def warn(self, message):
        ...

    @abstractmethod
    def error(self, message, error=None):
        ...


class NullLogger(BridgeLogger):
    """Discards all log messages."""

    def debug(self, message):
        pass

    def info(self, message):
        pass

    def warn(self, message):
        pass

    def error(self, message, error=None):
        pass


class ConsoleLogger(BridgeLogger):
    """Prints log messages to stdout."""

    def debug(self, message):
        print("[DEBUG] %s" % message)

    def info(self, message):
        print("[INFO] %s" % message)

    def warn(self, message):
        print("[WARN] %s" % message)

    def error(self, message, error=None):
        print("[ERROR] %s" % message)
        if error is not None:
            print("  Error: %s" % error)

File: dart/lib/src/browser_bridge.dart
/// Abstraction for browser message passing.
///
/// Implement this for any Dart web view (e.g. `webview_flutter`'s
/// `WebViewController`, `flutter_inappwebview`, a mock for testing). The Dart
/// equivalent of .NET's `IBrowserBridge`.
abstract class BrowserBridge {
  /// Whether the browser is ready for communication.
  bool get isReady;

  /// Callback invoked when a message is received from JavaScript.
  /// `BridgeHost` assigns this when it is constructed.
  set onMessage(void Function(String message)? callback);
  void Function(String message)? get onMessage;

  /// Send a message to JavaScript by executing a script.
  Future<void> executeScript(String script);

  /// Release any resources and detach from the underlying web view.
  void dispose();
}

/// Simple logging interface for `BridgeHost`.
abstract class BridgeLogger {
  void debug(String message);
  void info(String message);
  void warn(String message);
  void error(String message, [Object? error]);
}

/// Null logger that discards all messages (no-op).
class NullLogger implements BridgeLogger {
  const NullLogger();

  @override
  void debug(String message) {}
  @override
  void info(String message) {}
  @override
  void warn(String message) {}
  @override
  void error(String message, [Object? error]) {}
}

/// Logger that prints to stdout.
class ConsoleLogger implements BridgeLogger {
  const ConsoleLogger();

  @override
  void debug(String message) => print('[DEBUG] $message');
  @override
  void info(String message) => print('[INFO] $message');
  @override
  void warn(String message) => print('[WARN] $message');
  @override
  void error(String message, [Object? error]) {
    print('[ERROR] $message');
    if (error != null) print('  Error: $error');
  }
}
